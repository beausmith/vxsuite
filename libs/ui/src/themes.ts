/* istanbul ignore file */

import { Dictionary } from '@votingworks/types';
import {
  DEFAULT_FONT_SIZE,
  FONT_SIZES,
  LARGE_DISPLAY_FONT_SIZE,
} from './globals';

export interface FontSizeTheme {
  fontSize?: string;
}

export interface ColorTheme {
  background?: string;
  color?: string;
}

export interface Theme extends FontSizeTheme, ColorTheme {}

export const fontSizeTheme: Dictionary<FontSizeTheme> = {
  normal: {
    fontSize: `${FONT_SIZES[DEFAULT_FONT_SIZE]}px`,
  },
  medium: {
    fontSize: `${FONT_SIZES[2]}px`,
  },
  large: {
    fontSize: `${FONT_SIZES[LARGE_DISPLAY_FONT_SIZE]}px`,
  },
};

export const contrastTheme: Dictionary<ColorTheme> = {
  default: {
    color: '#263238',
    background: '#edeff0',
  },
  dark: {
    color: '#ffffff',
    background: '#455a64',
  },
};
