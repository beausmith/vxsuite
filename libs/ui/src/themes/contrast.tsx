import { ContrastSetting } from '@votingworks/types';

interface ContrastTheme {
  foreground: string;
  background: string;
}

export const contrastThemes: Record<ContrastSetting, ContrastTheme> = {
  black: {
    foreground: '#000000',
    background: '#ffffff',
  },
  white: {
    foreground: '#ffffff',
    background: '#000000',
  },
  grey: {
    foreground: '#ffffff',
    background: '#696969',
  },
};
export const defaultContrastTheme = contrastThemes['black'];
// const contrastThemeKeys = Object.keys(contrastThemes) as ContrastSetting[];
