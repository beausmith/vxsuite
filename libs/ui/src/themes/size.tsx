import { DefinedDictionary, SizeSetting, SizeTheme } from '@votingworks/types';

export const sizeThemes: DefinedDictionary<SizeTheme> = {
  S: {
    label: 'Small',
    text: 10,
    lineHeight: 1.4,
    h1: 1.35,
    h2: 1.15,
    h3: 1,
    h4: 0.8,
  },
  M: {
    label: 'Medium',
    text: 14,
    lineHeight: 1.4,
    h1: 1.35,
    h2: 1.15,
    h3: 1,
    h4: 0.8,
  },
  L: {
    label: 'Large',
    text: 18,
    lineHeight: 1.4,
    h1: 1.2,
    h2: 1.1,
    h3: 1,
    h4: 0.8,
  },
  XL: {
    label: 'Extra Large',
    text: 24,
    lineHeight: 1.2,
    h1: 1.2,
    h2: 1.1,
    h3: 1,
    h4: 0.8,
  },
};
export const sizeThemeKeys = Object.keys(sizeThemes) as SizeSetting[];
export const defaultSizeTheme = sizeThemes['M'];
export const extraLargeSizeTheme = sizeThemes['XL'];

const defaultPixelScreen: DefinedDictionary<number> = {
  // Default to Elo screen (1920 x 1080 FHD) for jest as it's screen is 0 x 0.
  // Could use the following in every test, but it would be redundant:
  //   jest.spyOn(window.screen, "width", "get").mockReturnValue(1080);
  //   jest.spyOn(window.screen, "height", "get").mockReturnValue(1920);
  width: window.screen.width || 1080,
  height: window.screen.height || 1920,
};

export const defaultPhysicalScreen: DefinedDictionary<number> = {
  // Elo 13" (E683595) 11.57" x 6.5"  (165.946413137424373ppi) = 166ppi
  // width: 6.5,
  // height: 11.57,
  // Elo 15" (E155645) 13.55" x 7.62" (141.697416974169742ppi) = 142ppi
  width: 7.62,
  height: 13.55,
};

export function screenPixelsPerInch(
  physical = defaultPhysicalScreen,
  pixel = defaultPixelScreen
): number {
  return Math.round(
    Math.hypot(pixel['width'], pixel['height']) /
      Math.hypot(physical['width'], physical['height'])
  );
}

export function cssFontSizePixels(points: number): number {
  return Math.round((screenPixelsPerInch() / 72) * points * 1.4); // 1.4 is a magic number to increase the font-size such that they letter "I" is the specified height in points.
}

export const cssFontSizeMinPixels = `${cssFontSizePixels(
  sizeThemes['S']['text']
)}px`;

export function screenInchesToPixels(
  inches: number,
  units = true,
  ppi = screenPixelsPerInch()
): number | string {
  const pixels = Math.round(ppi * inches);
  return units ? `${pixels}px` : pixels;
}
