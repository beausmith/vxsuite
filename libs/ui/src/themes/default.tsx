import { SizeTheme, ContrastTheme } from '@votingworks/types';
import { defaultContrastTheme } from './contrast';
import { defaultSizeTheme } from './size';

export interface Theme {
  size: SizeTheme;
  contrast: ContrastTheme;
}

export const defaultTheme: Theme = {
  size: defaultSizeTheme,
  contrast: defaultContrastTheme,
};
