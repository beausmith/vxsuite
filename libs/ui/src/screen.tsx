import styled from 'styled-components';
import { defaultTheme } from './themes/default';
import { cssFontSizePixels } from './themes/size';

interface Props {
  navRight?: boolean;
}

// The Screen component:
// - defines the foundational layout of the screen (flex-direction).
// - defines the base font-size of the page.
// - defined the background color and text color.

export const Screen = styled.div<Props>`
  display: flex;
  flex-direction: ${({ navRight }) => (navRight && 'row') || 'column'};
  background-color: ${({ theme }) => theme.contrast.background};
  height: 100%;
  overflow: auto;
  line-height: 1;
  color: ${({ theme }) => theme.contrast.foreground};
  font-size: ${({ theme }) => cssFontSizePixels(theme.size.text)}px;
  @media print {
    display: none;
  }
`;

Screen.defaultProps = {
  theme: defaultTheme,
};
