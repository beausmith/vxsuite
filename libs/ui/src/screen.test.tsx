import React from 'react';
import { render } from '@testing-library/react';

import { ThemeProvider } from 'styled-components';
import { Screen } from './screen';
import { contrastThemes } from './themes/contrast';
import { extraLargeSizeTheme } from './themes/size';

describe('renders Screen', () => {
  test('with defaults', () => {
    const { container } = render(<Screen>Screen</Screen>);
    const screen = container.firstChild;
    // critical styles
    expect(screen).toHaveStyleRule('height', '100%');
    expect(screen).toHaveStyleRule('display', 'none', {
      media: 'print',
    });
    // variable styles
    expect(screen).toHaveStyleRule('flex-direction', 'column');
    expect(screen).toHaveStyleRule('background-color', '#ffffff');
    expect(screen).toHaveStyleRule('color', '#000000');
    expect(screen).toHaveStyleRule('font-size', '39px');
  });

  test('with white contrast theme', () => {
    const { container } = render(
      <ThemeProvider
        theme={{
          size: extraLargeSizeTheme,
          contrast: contrastThemes['white'],
        }}
      >
        <Screen>Screen</Screen>
      </ThemeProvider>
    );
    const screen = container.firstChild;
    expect(screen).toHaveStyleRule('background-color', '#000000');
    expect(screen).toHaveStyleRule('font-size', '66px');
  });

  test('with right nav', () => {
    const { container } = render(<Screen navRight>Screen</Screen>);
    const screen = container.firstChild;
    expect(screen).toHaveStyleRule('flex-direction', 'row');
  });
});
