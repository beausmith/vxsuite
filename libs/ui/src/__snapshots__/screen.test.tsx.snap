// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Screen with defaults 1`] = `
<div
  class="sc-bdvvaa fgnmIK"
>
  <div>
    Hello
  </div>
   
  <div>
    World
  </div>
</div>
`;

exports[`renders Screen with with all non-default options 1`] = `
<div
  class="sc-bdvvaa fnFHGz"
>
  <div>
    Hello
  </div>
   
  <div>
    World
  </div>
</div>
`;
